### PR TITLE
feat(STONEINTG-538): refactor EnsureAllIntegrationTestPipelinesExist()

### PR DESCRIPTION
**Pull request by hongweiliu17**
* refactor the function to create integrationPipelinerun

Signed-off-by: Hongwei Liu <hongliu@redhat.com>
## Maintainers will complete the following section

- [x] Commit messages are descriptive enough ([hints](https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/))
- [x] Code coverage from testing does not decrease and new code is covered
- [x] [Controllers diagrams](https://github.com/redhat-appstudio/integration-service/tree/main/docs) are updated when PR changes controllers code  (if applicable)

### CODE REVIEW

File: controllers/snapshot/snapshot_adapter.go [hunk 1]
```diff
@@ -120,29 +120,14 @@ func (a *Adapter) EnsureAllIntegrationTestPipelinesExist() (controller.Operation
 					"integrationTestScenario.Name", integrationTestScenario.Name,
 					"len(integrationPipelineRuns)", len(*integrationPipelineRuns))
 			} else {
-				a.logger.Info("Creating new pipelinerun for integrationTestscenario",
-					"integrationTestScenario.Name", integrationTestScenario.Name)
 				pipelineRun, err := a.createIntegrationPipelineRun(a.application, &integrationTestScenario, a.snapshot)
-				if err != nil {
+				if pipelineRun == nil && err != nil {
 					a.logger.Error(err, "Failed to create pipelineRun for snapshot and scenario")
```
Comment (jsztuka) [inline controllers/snapshot/snapshot_adapter.go:125]: We may inform which pipelinerun failed as part as prompting error in the log.
Just out of curiosity, asking about the condition:  **pipelineRun == nil && err != nil**
What is the case when pipelinerun is not nil and error is not nil either? Is there such situation?

File: controllers/snapshot/snapshot_adapter.go [hunk 2]
```diff
@@ -554,13 +542,25 @@ func (a *Adapter) createIntegrationPipelineRun(application *applicationapiv1alph
...
+		return nil, fmt.Errorf("failed to call client.Create to create pipelineRun for snapshot %s: %w", snapshot.Name, err)
 	}
 
+	a.logger.LogAuditEvent("IntegrationTestscenario pipeline has been created", pipelineRun, h.LogActionAdd,
+		"integrationTestScenario.Name", integrationTestScenario.Name)
+	gitops.PrepareToRegisterIntegrationPipelineRun(a.snapshot)
+	if gitops.IsSnapshotNotStarted(a.snapshot) {
+		_, err := gitops.MarkSnapshotIntegrationStatusAsInProgress(a.client, a.context, a.snapshot, "Snapshot starts being tested by the integrationPipelineRun")
+		if err != nil {
+			a.logger.Error(err, "Failed to update integration status condition to in progress for snapshot")
```
Comment (jsztuka) [inline controllers/snapshot/snapshot_adapter.go:558]: Can we include snapshot iitself here? So its easier to search for it.

Comment (dirgim) [reply controllers/snapshot/snapshot_adapter.go:558]: I think that the Snapshot that's being reconciled will always be reported in the logs on account of the `r.Log.WithValues` call [here](https://github.com/redhat-appstudio/integration-service/blob/d302862916d13621967f34c14f5d91f46dac8d28/controllers/snapshot/snapshot_controller.go#L78).

File: controllers/snapshot/snapshot_adapter.go (hunk 1 above)
Comment (MartinBasti) [reply controllers/snapshot/snapshot_adapter.go:125]: I agree, if there is err != nil, we shouldn't care about return variable at all

Comment (hongweiliu17) [reply controllers/snapshot/snapshot_adapter.go:125]: 1. When the pipelineRun is not created successfully, I don't think we need to log it out.
2. When the pipelineRun is created successfully but it fails to update snapshot to `InProgress`, we just log some error but won't `return controller.RequeueWithError(err)`, see [here](https://github.com/redhat-appstudio/integration-service/blob/main/controllers/snapshot/snapshot_adapter.go#L138-L139).
    When we failed to set ControllerReference or create pipelineRun, we will return `nil, error` and then `return controller.RequeueWithError(err)`

File: controllers/snapshot/snapshot_adapter.go (hunk 2 above)
Comment (hongweiliu17) [reply controllers/snapshot/snapshot_adapter.go:558]: Thanks Kruno for reply. Yes.

File: controllers/snapshot/snapshot_adapter.go (hunk 1 above)
Comment (MartinBasti) [reply controllers/snapshot/snapshot_adapter.go:125]: 2. it still means that we need to check err != nil only in that condition, it cannot happen that we return error and the working pipeline at the same time

Comment (hongweiliu17) [reply controllers/snapshot/snapshot_adapter.go:125]: Good point, there was a check for `err != nil`, but I removed it. Updated. Thanks!

**Review by MartinBasti — APPROVED** [review]
Looks good to me, I'm not sure if it covers everything what Ryan would need in his epic, but this particular PR is good start, further improvements can be in new PRs.

**Review by dirgim — APPROVED** [review]
LGTM

**Comment by dirgim** [comment]
@14rcole can you also take a look if this aligns with the plans for [STONEINTG-463](https://issues.redhat.com/browse/STONEINTG-463)?

File: controllers/snapshot/snapshot_adapter.go [hunk 3]
```diff
@@ -120,29 +120,14 @@ func (a *Adapter) EnsureAllIntegrationTestPipelinesExist() (controller.Operation
...
 				pipelineRun, err := a.createIntegrationPipelineRun(a.application, &integrationTestScenario, a.snapshot)
 				if err != nil {
 					a.logger.Error(err, "Failed to create pipelineRun for snapshot and scenario")
 					return controller.RequeueWithError(err)
 				}
-				a.logger.LogAuditEvent("IntegrationTestscenario pipeline has been created", pipelineRun, h.LogActionAdd,
-					"integrationTestScenario.Name", integrationTestScenario.Name)
 				testStatuses.UpdateTestStatusIfChanged(
 					integrationTestScenario.Name, gitops.IntegrationTestStatusInProgress,
 					fmt.Sprintf("IntegrationTestScenario pipeline '%s' has been created", pipelineRun.Name))
```
Comment (14rcole) [inline controllers/snapshot/snapshot_adapter.go:130]: This is the only piece I'm not sure about.  Why isn't this also being moved to `createIntegrationPipelineRun()`? We would need to update the test status for the scenario in the case of a re-run.  Do you think this is just out of scope?

**Comment by 14rcole** [comment]
I had a question about another unmodified part of the code
```
if !reflect.ValueOf(integrationTestScenario.Spec.Environment).IsZero() {
				// the test pipeline for scenario needing an ephemeral environment will be handled in STONEINTG-333
				a.logger.Info("IntegrationTestScenario has environment defined, skipping creation of pipelinerun.", "IntegrationTestScenario", integrationTestScenario)
				continue
			}
```

Will `integrationTestScenario.Spec.Environment` always be defined for a scenario that has been run before?  Or do we need to check if the environment is defined in the case of a re-run?

File: controllers/snapshot/snapshot_adapter.go (hunk 3 above)
Comment (hongweiliu17) [reply controllers/snapshot/snapshot_adapter.go:130]: I had the similar question [here](https://redhat-internal.slack.com/archives/C034UFMUSV6/p1695279001912279), the status will be reported [here](https://github.com/redhat-appstudio/integration-service/blob/main/controllers/integrationpipeline/integrationpipeline_adapter.go#L194) as `Inprogress/TestPass/TestFailed` as well, I think it is reasonable to not to add a lot of code here when creating PLR in snapshot and binding controller.

**Comment by hongweiliu17** [comment]
> I had a question about another unmodified part of the code
> 
> ```
> if !reflect.ValueOf(integrationTestScenario.Spec.Environment).IsZero() {
> 	// the test pipeline for scenario needing an ephemeral environment will be handled in STONEINTG-333
> 	a.logger.Info("IntegrationTestScenario has environment defined, skipping creation of pipelinerun.", "IntegrationTestScenario", integrationTestScenario)
> 	continue
> }
> ```
> 
> Will `integrationTestScenario.Spec.Environment` always be defined for a scenario that has been run before? Or do we need to check if the environment is defined in the case of a re-run?



> I had a question about another unmodified part of the code
> 
> ```
> if !reflect.ValueOf(integrationTestScenario.Spec.Environment).IsZero() {
> 	// the test pipeline for scenario needing an ephemeral environment will be handled in STONEINTG-333
> 	a.logger.Info("IntegrationTestScenario has environment defined, skipping creation of pipelinerun.", "IntegrationTestScenario", integrationTestScenario)
> 	continue
> }
> ```
> 
> Will `integrationTestScenario.Spec.Environment` always be defined for a scenario that has been run before? Or do we need to check if the environment is defined in the case of a re-run?

I guess the rerun logic would be like this: If environment is defined in ITS, call `CreateCopyOfEnvironment()` which is being refactored in [STONEINTG-539](https://issues.redhat.com/browse/STONEINTG-539), otherwise, call `CreateIntegrationPLR()` defined here.

**Comment by 14rcole** [comment]
> I guess the rerun logic would be like this: If environment is defined in ITS, call CreateCopyOfEnvironment() which is being refactored in [STONEINTG-539](https://issues.redhat.com/browse/STONEINTG-539), otherwise, call CreateIntegrationPLR() defined here. Please correct me if my understanding is incorrect.

That sounds right

**Review by Josh-Everett — APPROVED** [review]

**Comment by hongweiliu17** [comment]
> > I guess the rerun logic would be like this: If environment is defined in ITS, call CreateCopyOfEnvironment() which is being refactored in [STONEINTG-539](https://issues.redhat.com/browse/STONEINTG-539), otherwise, call CreateIntegrationPLR() defined here. Please correct me if my understanding is incorrect.
> 
> That sounds right

@14rcole Thanks for checking. So does this PR look good to you?

**Review by dheerajodha — APPROVED** [review]

**Comment by sonam1412** [comment]
> > > I guess the rerun logic would be like this: If environment is defined in ITS, call CreateCopyOfEnvironment() which is being refactored in [STONEINTG-539](https://issues.redhat.com/browse/STONEINTG-539), otherwise, call CreateIntegrationPLR() defined here. Please correct me if my understanding is incorrect.
> > 
> > 
> > That sounds right
> 
> @14rcole Thanks for checking. So does this PR look good to you?

@14rcole Can you please confirm if this looks good to you and can be merged  ?

**Review by 14rcole — APPROVED** [review]

**Pull request merged**